compiler/IREmitter/Payload/patches/vm_insnhelper.c

```diff
@@ -481,3 +484,232 @@ void sorbet_throwReturn(rb_execution_context_t *ec, VALUE retval) {
+        }
+        rb_set_errinfo(previousException);
+    } else {
+        // If we get here, setjmp has returned a non-zero value.
+
+        // This case is the "obvious" case: something in the body threw; we need to handle
+        // exceptions directly here.
+        if (state == RunningBody) {
+            // We're handling things very similarly to rb_rescue2/rb_vrescue2:
+            // <add-link-here>
+            //
+            // The significant difference from that function is that we're handling all
+            // the non-local exits directly.
+
+            // Whatever kind of non-local exit we have, we need to make sure that
+            // the Ruby control frame for the handler (or the ensure) we're going
+            // to run lives directly under whatever frame we started this process with.
+            rb_vm_rewind_cfp(ec, cfp);
+
+            if (nleType == TAG_RAISE) {
+                // rb_rescue2/rb_vrescue2 would check ec->errinfo here to determine if it
+                // was the "right" kind of error.  Sorbet has already generated code to check
+                // the exception value for us, we can dispatch directly to the handlers here,
+                // which avoids a little bit of overhead.  But we do need to tell the
+                // handlers what the exception value *is*.
+                bodyException = ec->errinfo;
+            } else {
+                // Any other kind of non-local exit will skip the rescue/else handlers.
+                goto execute_ensure;
+            }
+
+            // Any exception that got thrown needs to be set for the handler.
+            sorbet_writeLocal(cfp, exceptionValueIndex, exceptionValueLevel, bodyException);
+
+            ExceptionFFIType handler = (bodyException != Qnil) ? handlers : elseClause;
+
+            // Indicate that we are running the appropriate handler.
+            nleType = TAG_NONE;
+            state = RunningHandlers;
+
+            executionResult = handler(pc, methodClosure, cfp);
+
+            // If execution has reached this point, the handler did not throw any
+            // kind of non-local exit.
+            if (bodyException != Qnil && executionResult == retrySingleton) {
+                goto execute_body;
+            }
+        } else {
+            // This case is the non-obvious case: something in the handler we were executing
+            // threw.  We need to clean up and execute the ensure.
+
+            rb_vm_rewind_cfp(ec, cfp);
+
+            // TODO(froydnj): this is where we would handle TAG_RETRY if we were implementing
+            // `retry` in terms of Ruby's non-local exit handling rather than our current
+            // retry singleton mechanism.
+
+            if (nleType == TAG_RAISE) {
+                handlerException = ec->errinfo;
+                sorbet_writeLocal(cfp, exceptionValueIndex, exceptionValueLevel, handlerException);
+            } else {
+                goto execute_ensure;
+            }
```

Comment:
Again, I'd rather that we invert this so that only the `goto` is guarded.
```suggestion
            if (nleType != TAG_RAISE) {
                goto execute_ensure;
            }
            
            handlerException = ec->errinfo;
            sorbet_writeLocal(cfp, exceptionValueIndex, exceptionValueLevel, handlerException);
```